src/pointer-analysis/value_set_dereference.cpp

```diff
@@ -643,11 +643,12 @@ value_set_dereferencet::valuet value_set_dereferencet::build_reference_to(
       }
     }
 
-    // try to build a member/index expression - do not use byte_extract
+    // Try to build a member/index expression instead of using byte_extract;
+    // don't recurse into aggregate types as the caller's expression may be
+    // within an index expression, i.e., a further offset beyond o.offset() may
+    // get added on to it.
     auto subexpr = get_subexpression_at_offset(
-      root_object, o.offset(), dereference_type, ns);
-    if(subexpr.has_value())
-      simplify(subexpr.value(), ns);
+      root_object, o.offset(), dereference_type, false, ns);
```

Comment:
Can you summarize what conditions hold when reaching this line  (i.e. what forces flipping `recurse_into_arrays` to false ?)

Reply:
From the comment right above it rather seems like if recurse_into_arrays is false then you try to build a field access expression, but when it is true you try to build a byte-extract that can hit the array anywhere ?

Reply:
Thank you for asking all those questions, they just highlighted that the initial change was a bit dubious. Now much simpler and clearer.